README.md
# CreatiSign Generator
Arc Sign Codes generator for [Creativerse](https://store.steampowered.com/app/280790/Creativerse/), free to play on Steam.

This tool allows for converting images into codes that can be pasted into Arc Signs - that is, importing images inside Creativerse.

Live latest version on [my website](http://entuland.com/creatisign)

Discussion about CreatiSign Generator on the [Steam Forums](https://steamcommunity.com/app/280790/discussions/3/1479857071263583178/)

If you like my contributions you may consider reading http://entuland.com/en/support-entuland


changelog.txt
CreatiSign Generator
Proudly made in Sardinia (Italy) by Entuland (entuland@gmail.com)

Changelog:

V3.4.2:
	- removed the warning about the default size and main spacing (didn't really hit the repo, only affected the live version IIRC)
	- hopefully fixed the code that optimizes transparent pixels

V3.4:
	- added import/export facilities for settings
	- got rid of all useless transparent pixels to gain some resolution in certain cases
	- refactored the UI a little
	- altered default size and main spacing to improve quality at a distance

V3.3:
	- added whole new section for source editing: cropping, rotation, mirroring, source-width control, splitting image into multiple sections to be processed separately
	- moved most of the info into dedicated overlay
	- added a bit of instructions in the above mentioned overlay
	- moved reset settings button to the left (cause it only applies to the left-hand-side settings, i.e. the generation settings)
	- removed the obnoxious "select all on focus" for regular textboxes, only applies to yellow textareas now

V3.2:
	- free placement of arc signs now gets restored properly across page reloads
	- arc signs get marked in red after the processing only if they actually overflow any limit
	- the image preview in the placement preview area now gets resized in height properly after execution
	- replaced all the images with data URIs to ensure the page looks right even when downloaded and run from a local copy

V3.1.BETA:
	- rebuilt preview placement area completely, allowing for free placement of single arc signs
	- refactored main slicing algorithm to avoid locking the browser and to show a progress of the actual image processing

V3:
	- reorganized interface grouping settings in different modes (basic, medium, advanced, insane)
	- added settings memory (survives page refreshes) and a settings reset button
	- added vertical error correction (to account for slight overlaps from slice to slice)
	- added free-text incipit for each code
	- added <indent> tag
	- added <align> tag
	- added icons to most settings
	- after execution, the height of the image in the placement preview area survives the change of settings to let you tweak the <offset> value
	- removed pattern recipes (creating patterns is still possible just like in previous version, I've only removed the recipes: if you were smart enough to understand them you're smart enough to create them fiddling with the settings on your own, besides no one seems to have used them at all so far)

V2.6:
	- fixed overlapping, duplicated row across slides
	- improved slide placement (they were slightly off even before the duplicated row)
	- fixed line count in each textarea
	- improved preview of sign placement
	- added preview of graphic extension

V2.5:
	- selecting / dragging an image shows its preview immediately
	- added ability to arrange arc signs in different ways
	- added the preview of both the signs arrangement and the graphics placement related to the arc signs (hover it to enlarge it)
	- the graphic now starts exactly at the chosen offset and grows towards the bottom, not somewhat higher than the chosen offset as it happened in previous versions
	- removed the ability to omit offset and zoffset tags by setting them to zero (the offset defaulting to about 0.55 when omitted is misleading in first place, and keeping them always in the code makes it easier to tweak them in the game)
	- added checkered background to the source and result previews to highlight transparent areas
	- the result preview now shows the gamma correction and grayscale effects
	- the textareas get clearly marked in red if any limit gets exceeded
	- added a "Working..." cover to show when the script is running
	- changed the default value of offset from 0 to 2, to make it stand a bit higher than the first arc sign
	- changed the default value of zoffset from 0.51 to 0 to have it stand exactly in the middle of the column
	- refactored the generation algorithm to optimize the slices usage (there won't be any wasted signs using just 2 or 3 K chars, perhaps just the last one depending on the settings); this got also rid of the bug that trimmed some lines at the end of the source image
	- refactored the whole script into separate files to ease the development

V2.4:
	- added slicing feature
	- added grayscale convertion option
	- added offset and zoffset controls
	- added font & character controls
	- added some recipes to create simple patterns
	- major change in spacing tweak: its sign has been reversed and now the <size> value remains untouched, it's the <mspace> and <line-height> values that change depending on the tweak
	- restyled the page a bit and tidied up code a little

V2.3: fixed extremely silly bug in alpha handling

V2.2: added gamma correction, fixed a bug that created artifacts at the border of the image in some cases

V2.1: converted line endings from "\n" to "\r\n"

V2: uses alternate font and periods to draw fully opaque pixels, added options to tweak the tag values, added autosize feature, added short hex values feature

V1: first version using <mark> technique

js/edit.js
'use strict';

var edit = {
	
	init: function() {
		edit.initControls();
		edit.initEvents();
	},
	
	exportSettings: function() {
		return {
			settingsToggle: edit.settingsToggle.checked,
			forceWidth: edit.forceWidth.checked,
			sectionWidth: edit.sectionWidth.value,
			splitCols: edit.splitCols.value,
			splitRows: edit.splitRows.value,
			cropTop: edit.cropTop.value,
			cropLeft: edit.cropLeft.value,
			cropRight: edit.cropRight.value,
			cropBottom: edit.cropBottom.value,
			transparents: edit.transparents.value,
		};
	},
	
	importSettings: function(settings) {
		edit.settingsToggle.checked = settings.settingsToggle;
		edit.forceWidth.checked = settings.forceWidth;
		edit.sectionWidth.value = settings.sectionWidth;
		edit.splitCols.value = settings.splitCols;
		edit.splitRows.value = settings.splitRows;
		edit.cropTop.value = settings.cropTop;
		edit.cropLeft.value = settings.cropLeft;
		edit.cropRight.value = settings.cropRight;
		edit.cropBottom.value = settings.cropBottom;
		edit.transparents.value = settings.transparents;
		edit.settingsToggleToggled();
		edit.forceWidthToggled();
		edit.updateCropping();
	},
	
	initControls: function() {
		edit.colors = [];
		edit.settingsToggle = document.querySelector('#showsourcesettings');
		edit.sourceSettingsTable = document.querySelector('#sourcesettings');
		edit.transparents = document.querySelector('#transparents');
		edit.forceWidth = document.querySelector('#forcewidth');
		edit.sectionWidth = document.querySelector('#sectionwidth');
		edit.splitCols = document.querySelector('#splitcolumns');
		edit.splitRows = document.querySelector('#splitrows');
		edit.cropAll = document.querySelector('#cropall');
		edit.cropTop = document.querySelector('#croptop');
		edit.cropRight = document.querySelector('#cropright');
		edit.cropBottom = document.querySelector('#cropbottom');
		edit.cropLeft = document.querySelector('#cropleft');
		edit.cropMask = document.querySelector('#cropmask');
		edit.cropMouse = document.querySelector('#cropmouse');
		edit.canvas = document.querySelector('#editcanvas');
		edit.rotateLeft = document.querySelector('#rotateleft');
		edit.rotateRight = document.querySelector('#rotateright');
		edit.mirrorVertical = document.querySelector('#mirrorvertical');
		edit.mirrorHorizontal = document.querySelector('#mirrorhorizontal');
		edit.ctx = edit.canvas.getContext('2d');
		edit.sectionNotice = document.querySelector('#sectionnotice');
	},
	
	initEvents: function() {
		edit.forceWidth.addEventListener('change', edit.forceWidthToggled);
		edit.splitRows.addEventListener('input', edit.updateCropping);
		edit.splitCols.addEventListener('input', edit.updateCropping);
		edit.cropMouse.addEventListener('mousedown', edit.maskMouseDown);
		edit.cropMouse.addEventListener('mousemove', edit.maskMouseMove);
		document.addEventListener('mouseup', edit.maskMouseUp);
		edit.cropAll.addEventListener('input', edit.cropAllChange);
		edit.cropTop.addEventListener('input', edit.updateCropping);
		edit.cropRight.addEventListener('input', edit.updateCropping);
		edit.cropLeft.addEventListener('input', edit.updateCropping);
		edit.cropBottom.addEventListener('input', edit.updateCropping);
		edit.settingsToggle.addEventListener('click', edit.settingsToggleToggled);
		edit.rotateLeft.addEventListener('click', function() {
			edit.rotate('left');
		});
		edit.rotateRight.addEventListener('click', function() {
			edit.rotate('right');
		});
		edit.mirrorHorizontal.addEventListener('click', function() {
			edit.mirror('horizontal');
		});
		edit.mirrorVertical.addEventListener('click', function() {
			edit.mirror('vertical');
		});
	},
	
	filterColor: function(hex) {
		if(!edit.colors.length) {
			return hex;
		}
		if(edit.colors.indexOf(hex.replace('#', '')) !== -1) {
			return '#0000';
		}
		return hex;
	},
	
	prepareColors: function() {
		var remove = edit.transparents.value.toUpperCase();		
		edit.colors = remove.replace(/[^0-9A-F]+/g, ' ').replace(/\s+/g, ' ').split(' ');
	},
	
	settingsToggleToggled: function() {
		edit.sourceSettingsTable.style.display = edit.settingsToggle.checked ? "grid" : "none";
	},
	
	forceWidthToggled: function() {
		if(edit.forceWidth.checked) {
			edit.sectionWidth.parentNode.style.visibility = 'visible';
		} else {
			edit.sectionWidth.parentNode.style.visibility = 'hidden';
		}
	},
	
	maskMouseDown: function(e) {
		var rect = inter.getRect(edit.cropMouse);
		edit.maskStart = {
			x: e.clientX - rect.x,
			y: e.clientY - rect.y,
		};
	},
	
	maskMouseMove: function(e) {
		if(!edit.maskStart) {
			return;
		}
		var a = edit.maskStart;
		var rect = inter.getRect(edit.cropMouse);
		var b = {
			x: e.clientX - rect.x,
			y: e.clientY - rect.y,
		};
		edit.maskToCrop(a, b);
	},
	
	maskMouseUp: function(e) {
		edit.maskStart = false;
	},
	
	maskToCrop: function(a, b) {
		var minX = Math.min(a.x, b.x);
		var maxX = Math.max(a.x, b.x);
		var minY = Math.min(a.y, b.y);
		var maxY = Math.max(a.y, b.y);
		
		if(maxX - minX < 5 || maxY - minY < 5) {
			return;
		}
		
		var source = edit.canvas;
		var prev = inter.els.sourcePreview;
		
		var ratio = source.width / prev.clientWidth;

		edit.cropLeft.value = Math.round(minX * ratio);
		edit.cropTop.value = Math.round(minY * ratio);
		edit.cropRight.value = Math.round((prev.clientWidth - maxX) * ratio);
		edit.cropBottom.value = Math.round((prev.clientHeight - maxY) * ratio);
		
		window.ddd = {
			a: a,
			b: b,
			ratio: ratio,
			prev: prev,
			source: source,
		};
		
		edit.updateCropping();
	},
	
	setSource: function(source) {
		edit.canvas.width = source.width;
		edit.canvas.height = source.height;
		edit.ctx.drawImage(source, 0, 0);
		setTimeout(edit.updateCropping, 100);
	},
	
	getSource: function() {
		var source = document.createElement('canvas');
		var ratio = 1;
		
		source.width = (edit.canvas.width - edit.cropLeft.value - edit.cropRight.value) / edit.splitCols.value;
		source.height = (edit.canvas.height - edit.cropTop.value - edit.cropBottom.value) / edit.splitRows.value;
		
		if(edit.forceWidth.checked) {
			ratio = parseFloat(edit.sectionWidth.value) / source.width;
		}
		source.width *= ratio;
		source.height *= ratio;
		
		var ctx = source.getContext('2d');
		// ctx.imageSmoothingEnabled = false;
		// ctx.translate(0.5, 0.5);
		
		var col = edit.chosenSection % edit.splitCols.value;
		var row = Math.floor(edit.chosenSection / edit.splitCols.value);
		
		var left = -source.width * col - edit.cropLeft.value * ratio;
		var top = -source.height * row - edit.cropTop.value * ratio;
		ctx.drawImage(edit.canvas, left, top, edit.canvas.width * ratio, edit.canvas.height * ratio);
		
		edit.prepareColors();
		
		return source;
	},
	
	updateSectionNotice: function() {
		edit.sectionNotice.innerText = "You're working on section #" + (edit.chosenSection + 1) + " of " + edit.sections.length;
	},
	
	updatePreview: function() {
		inter.els.sourcePreview.src = edit.canvas.toDataURL();
		setTimeout(edit.updateCropping, 100);
	},
	
	cropAllChange: function() {
		edit.cropTop.value 
			= edit.cropRight.value
			= edit.cropBottom.value
			= edit.cropLeft.value
			= edit.cropAll.value;
		edit.updateCropping();
	},
	
	updateCropping: function() {
		var source = edit.canvas;
		var prev = inter.els.sourcePreview;
		var mask = edit.cropMask;
		var mouse = edit.cropMouse;
		
		mask.innerHTML = '';
		
		var ratio = prev.clientWidth / source.width;
		
		mouse.style.display = 'block';
		mouse.style.width = prev.clientWidth + 'px';
		mouse.style.height = prev.clientHeight + 'px';
		
		mask.style.borderWidth =
			Math.round(edit.cropTop.value * ratio) + "px " + 
			Math.round(edit.cropRight.value * ratio) + "px " + 
			Math.round(edit.cropBottom.value * ratio) + "px " + 
			Math.round(edit.cropLeft.value * ratio) + "px";
		
		var total = edit.splitCols.value * edit.splitRows.value;
		
		edit.sections = [];
		
		for(var i = 0; i < total; ++i) {
			var section = document.createElement('div');
			section.classList.add('splitsection');
			mask.appendChild(section);
			section.dataset.index = i;
			if(i === 0) {
				section.classList.add('selected');
			}
			section.style.width = mask.clientWidth / edit.splitCols.value + 'px';
			section.style.height = mask.clientHeight / edit.splitRows.value + 'px';
			section.addEventListener('click', edit.chooseSection);
			section.setAttribute('title', 'Click to select, click and drag for manual cropping');
			edit.sections.push(section);
		}
		edit.chosenSection = 0;
		edit.updateSectionNotice();
	},
	
	chooseSection: function() {
		edit.chosenSection = parseInt(this.dataset.index);
		for(var i = 0; i < edit.sections.length; ++i) {
			var section = edit.sections[i];
			if(i === edit.chosenSection) {
				section.classList.add('selected');
			} else {
				section.classList.remove('selected');
			}
		}
		edit.updateSectionNotice();
	},
	
	rotate: function(direction) {
		
		var radians = Math.PI / 2;
		if(direction == 'left') {
			radians = -radians;
		}
		var canvas = document.createElement('canvas');
		canvas.width = edit.canvas.width;
		canvas.height = edit.canvas.height;

		var ctx = canvas.getContext('2d');
		ctx.drawImage(edit.canvas, 0, 0);

		var pass = edit.canvas.width;
		edit.canvas.width = edit.canvas.height;
		edit.canvas.height = pass;
		
		edit.ctx.clearRect(0, 0, edit.canvas.width, canvas.height);
		edit.ctx.save();
		edit.ctx.translate(edit.canvas.width / 2, edit.canvas.height / 2);
		edit.ctx.rotate(radians);
		edit.ctx.drawImage(canvas, -edit.canvas.height / 2, -edit.canvas.width / 2);
		edit.ctx.restore();
		edit.updatePreview();
	},
	
	mirror: function(direction) {
		edit.ctx.save();
		var canvas = document.createElement('canvas');
		canvas.width = edit.canvas.width;
		canvas.height = edit.canvas.height;

		var ctx = canvas.getContext('2d');

		if(direction == 'horizontal') {
			ctx.scale(-1, 1);
			ctx.drawImage(edit.canvas, 0, 0, -canvas.width, canvas.height); 
		} else {
			ctx.scale(1, -1);
			ctx.drawImage(edit.canvas, 0, 0, canvas.width, -canvas.height); 
		}
		
		edit.ctx.clearRect(0, 0, edit.canvas.width, canvas.height);
		edit.ctx.drawImage(canvas, 0, 0);
		edit.ctx.restore();
		edit.updatePreview();
	},

};

js/color.js
'use strict';

var color = {
	
	shortHex: function(hex) {
		if(hex % 17 > 7) {
			hex = 17 + hex - hex % 17;
		} else {
			hex = hex - hex % 17;
		}
		return ('' + hex.toString(16)).slice(0, 1);
	},

	expandedHex: function(hex) {
		var result = '#';
		for(var i = 1; i < hex.length; ++i) {
			result += hex[i] + hex[i];
		}
		return result;
	},

	rgbaToHex: function(rgba) {
		// for some reason weird numbers are arriving here, rounding them to play safe for the time being
		rgba.r = Math.round(rgba.r);
		rgba.g = Math.round(rgba.g);
		rgba.b = Math.round(rgba.b);
		rgba.a = Math.round(rgba.a);
		if (rgba.r > 255 || rgba.g > 255 || rgba.b > 255 || rgba.a > 255) {
			throw 'Invalid color component';
		}
		rgba = color.alphaCheck(rgba);
		var result = '#';		
		result += rgba.r.toString(16).padStart(2, "0");
		result += rgba.g.toString(16).padStart(2, "0");
		result += rgba.b.toString(16).padStart(2, "0");
		if(rgba.a < 255) {
			result += rgba.a.toString(16).padStart(2, "0");
		}
		return result.toUpperCase();
	},

	rgbaToHexShort: function(rgba) {
		rgba = color.alphaCheck(rgba);
		var result = '';
		rgba.r = color.shortHex(rgba.r);
		rgba.g = color.shortHex(rgba.g);
		rgba.b = color.shortHex(rgba.b);
		rgba.a = color.shortHex(rgba.a);
		if(rgba.a !== 'f') {
			return ("#" + rgba.r + rgba.g + rgba.b + rgba.a).toUpperCase();
		}
		return ("#" + rgba.r + rgba.g + rgba.b).toUpperCase();
	},
	
	hexToRgba: function(hex) {
		var trimmed = hex.replace(/^#/, '');
		var rgba = {
			r: 0,
			g: 0,
			b: 0,
			a: 255,
		}
		if([3, 4, 6, 8].indexOf(trimmed.length) == -1) {
			return rgba;
		}
		var expanded = trimmed;
		if(trimmed.length < 6) {
			expanded = '';
			for(var i = 0; i < trimmed.length; ++i) {
				expanded += '' + trimmed[i] + trimmed[i];
			}
		}
		rgba.r = parseInt(expanded.substr(0, 2), 16);
		rgba.g = parseInt(expanded.substr(2, 2), 16);
		rgba.b = parseInt(expanded.substr(4, 2), 16);
		if(expanded.length == 8) {			
			rgba.a = parseInt(expanded.substr(6, 2), 16);
		}
		return color.alphaCheck(rgba);
	},
	
	alphaCheck: function(rgba) {
		if(rgba.a == 0) {
			rgba.r = rgba.g = rgba.b = rgba.a;
		}
		return rgba;
	}
	
};



js/settings.js
'use strict';

var creatisign_settings_version = '3.4.2';

var settings = [

	// containers

	{
		tagname: 'div',
		attributes: {
			'id': 'basicSettings',
			'class': 'settingsContainer',
		},
		noWrapper: true,
		html: '<h2>Generation Settings</h2><small>Basic Settings</small>',
		parent: '#interface',
	},
	{
		tagname: 'div',
		attributes: {
			'id': 'mediumSettings',
			'class': 'settingsContainer',
		},
		noWrapper: true,
		html: '<small>Medium Settings</small>',
		parent: '#interface',
	},
	{
		tagname: 'div',
		attributes: {
			'id': 'advancedSettings',
			'class': 'settingsContainer',
		},
		noWrapper: true,
		html: '<small>Advanced Settings</small>',
		parent: '#interface',
	},

	{
		tagname: 'div',
		attributes: {
			'id': 'insaneSettings',
			'class': 'settingsContainer',
		},
		noWrapper: true,
		html: '<small>Insane Settings</small>',
		parent: '#interface',
	},
	
	// basic settings
	
	{
		tagname: 'select',
		attributes: {
			id: 'settingsMode',
			title: 'All settings are active and will remember their values regardless of being hidden or not.',
		},
		label: 'Settings mode',
		options: {
			'1': 'Basic',
			'2': 'Medium',
			'3': 'Advanced',
			'4': 'Insane',
		},
		parent: 'basicSettings',
		label: 'Settings Mode',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'shortCodes',
			type: 'checkbox',
			title: 'Use short color codes to help reducing the code size.\nThis setting can alter the colors of your image.\nShort color codes will be used anyway if they match the long color codes.',
		},
		icon: imgsrc['shortcodes.png'],
		label: 'Short codes',
		parent: 'basicSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'grayScale',
			type: 'checkbox',
			title: 'Convert the image to shades of gray.',
		},
		icon: imgsrc['grayscale.png'],
		label: 'Grayscale',
		parent: 'basicSettings',
	},


	
	// medium settings
	
	{
		tagname: 'input',
		attributes: {
			id: 'slices',
			type: 'number',
			min: '1',
			max: '64',
			step: '1',
			value: '1',
			title: 'Range 1 ~ 64\nNumber of horizontal slices / arc signs.\nLarger values provide better resolution.',
		},
		icon: imgsrc['slices.png'],
		label: 'Slices',
		parent: 'mediumSettings',
	},
	{
		tagname: 'select',
		attributes: {
			id: 'disposition',
			title: 'V and H dispositions are limited to 17 slices.\nYou need an <offset> of -8 to use 17 V slices.\nYou need a <zoffset> of -8 to use 17 H slices\nThe GRID disposition allows up to 64 slices\nAnytime you move signs manually you go into Free Placement mode and will need to add / remove signs one by one in the placement interface (bottom right).',
		},
		options: {
			grid: 'GRID (Horizontal + Vertical)',
			gridRight: 'GRID RIGHT',
			gridBottom: 'GRID BOTTOM',
			horizontal: 'H (Horizontal)',
			horizontalCentered: 'HC (Horizontal Centered)',
			vertical: 'V (Vertical)',
			verticalCentered: 'VC (Vertical Centered)',
			free: 'Free placement',
		},
		icon: imgsrc['disposition.png'],
		label: 'Sign disposition',
		parent: 'mediumSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'size',
			type: 'number',
			value: '2',
			step: '0.1',
			title: 'Change this value to "zoom" your image in or out.\nA pixel of size 1 will have a width slightly smaller than 1% of a block\'s width.\nActual perceived size depends on spacing, font and characters.',
		},
		unit: 'points',
		icon: imgsrc['size.png'],
		label: '<size> tag',
		parent: 'mediumSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'offset',
			type: 'number',
			step: '0.01',
			min: '-8',
			max: '8',
			value: '2',
			title: 'Defaults to about 0.55 in the game when omitted.\nPositive values bring upwards, negative values bring downwards.',
		},
		label: '<offset> tag',
		unit: 'blocks',
		icon: imgsrc['offset.png'],
		parent: 'mediumSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'zoffset',
			type: 'number',
			step: '0.01',
			min: '-8',
			max: '8',
			value: '0',
			title: 'Defaults to zero in the game when omitted and results in the graphic placed on the vertical of the arc sign.\nPositive values bring forward, negative values bring backwards.\nSticking images to the side of a full block requires values near a half unit (for instance, 0.51, 1.5, -2.49 and so forth, depends on the desired setup).',
		},
		unit: 'blocks',
		icon: imgsrc['zoffset.png'],
		label: '<zoffset> tag',
		parent: 'mediumSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'indent',
			type: 'number',
			value: '0',
			step: '0.01',
			title: 'Indent is a horizontal spacing applied to the beginning of each line.\nNegative values bring leftwards, positive values bring rightwards.\nDouble the values for CENTER alignment.\nThis setting is pretty much useless with RIGHT alignment.',
		},
		unit: 'blocks',
		icon: imgsrc['indent.png'],
		label: '<indent> tag',
		parent: 'mediumSettings',
	},
	
	
	// advanced settings

	{
		tagname: 'input',
		attributes: {
			id: 'autoSize',
			type: 'checkbox',
			checked: 'checked',
			title: 'Resize the image to fit it into the available slices.\nThis option may upscale and blur the image.',
		},
		label: 'Autosize',
		icon: imgsrc['autosize.png'],
		parent: 'advancedSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'width',
			type: 'number',
			min: '0',
			max: '8',
			value: '8',
			step: '0.1',
			title: 'Maximum amount of blocks the sign is allowed to grow horizontally before wrapping lines.\nThis space is centered on the middle of the arc sign.',
		},
		unit: 'blocks',
		label: '<width> tag',
		icon: imgsrc['width.png'],
		parent: 'advancedSettings',
	},
	{
		tagname: 'select',
		attributes: {
			id: 'align',
			title: 'LEFT and RIGHT align refer to the limits set by the <width> tag.',
		},
		options: {
			'': 'center',
			'left': 'left',
			'right': 'right',
		},
		label: '<align> tag',
		icon: imgsrc['align.png'],
		parent: 'advancedSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'decimals',
			type: 'number',
			value: '3',
			min: '0',
			title: 'Number of decimals for number rounding.',
		},
		icon: imgsrc['decimals.png'],
		label: 'Decimals',
		parent: 'advancedSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'gammaCorrection',
			type: 'number',
			min: '0.01',
			max: '7.99',
			value: '0.5',
			step: '0.01',
			title: 'Range 0.01 ~ 7.99, preset to 0.5 to work around the current deviation of RGB shades.\nSet this to 1 to disable gamma correction altogether.',
		},
		icon: imgsrc['gammacorrection.png'],
		label: 'Gamma Correction',
		parent: 'advancedSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'alphaCorrection',
			type: 'number',
			min: '0.01',
			max: '7.99',
			value: '1',
			step: '0.01',
			title: 'Range 0.01 ~ 7.99, defaults to 1 (disabled), only useful for art heavily relying on alpha gradients.',
		},
		icon: imgsrc['alphacorrection.png'],
		label: 'Alpha Correction',
		parent: 'advancedSettings',
	},
	
	// insane settings
	{
		tagname: 'input',
		attributes: {
			id: 'mainSpacing',
			type: 'number',
			value: '-10',	
			title: 'Values higher than -10% (such as -5%) start to show the spacing grid between pixels.\nThe higher the value, the larger the spacing.\nValues lower than 10% cause overlapping but give higher quality from a distance.',
		},
		unit: '%',
		icon: imgsrc['mainspacing.png'],
		label: 'Main spacing',
		parent: 'insaneSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'verticalSpacing',
			type: 'number',
			value: '100',
			title: '100% gives equal spacing in both directions, 200% doubles the vertical spacing, 50% halves it and so forth.',
		},
		unit: '%',
		icon: imgsrc['verticalspacing.png'],
		label: 'Vertical spacing',
		parent: 'insaneSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'vError',
			type: 'number',
			value: '1',
			title: 'Correction applied to the <offset> when multiplying <line-height> by the total number of lines in previous slices.\nOnly change this if you intimately know what this settings does and why.',
		},
		unit: '%',
		label: 'Vertical Error Correction',
		parent: 'insaneSettings',
	},
	{
		tagname: 'select',
		attributes: {
			id: 'fontName',
			title: 'The "Arial SDF" font provides fully opaque characters without any outline.\nThe "default" font is the semiopaque, outlined font used normally by all signs.\nDidn\'t really fiddle with the other fonts, use them at your own risk.',
		},
		options: {
			'Arial SDF': 'Arial SDF',
			'default': 'Default',
			'LiberationSans SDF': 'LiberationSans SDF',
			'SourceHanSerifSC-Bold SDF': 'SourceHanSerifSC-Bold SDF',
			'SourceHanSerifSC-Regular SDF': 'SourceHanSerifSC-Regular SDF',
		},
		label: 'Font name',
		parent: 'insaneSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'characters',
			type: 'text',
			value: '.',
			title: 'Defaults to a single "period / full stop" character.\nWhatever you type here will be repeated for each pixel of the image.',
		},
		label: 'Character(s)',
		parent: 'insaneSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'charLimit',
			type: 'number',
			value: '10000',
			title: 'Maximum amount of characters allowed in a single sign.\nThis is only used to decide how to slice the image in this web tool, the actual limit is imposed by the game.',
		},
		label: 'Character limit',
		parent: 'insaneSettings',
	},
	{
		tagname: 'input',
		attributes: {
			id: 'freeIncipit',
			type: 'text',
			value: '',
			title: 'Free text to be inserted at the beginning of each textarea.\nIt\'s meant for custom tags, not for readable text.',
		},
		label: 'Free text prefix',
		parent: 'insaneSettings',
	},

	{
		tagname: 'input',
		attributes: {
			id: '',
			type: '',
			title: '',
		},
		label: '',
		parent: '',
	},
];

